app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java

```diff
@@ -0,0 +1,174 @@
+            if (filename != null && new File(filename).exists()) {
+                files.add(filename);
+            }
+        }
+        return files;
+    }
+
+    protected static void removeFiles(List<String> filesToRemove) {
+        for (String filename : filesToRemove) {
+            File datafile = new File(filename);
+            datafile.delete();
+        }
+    }
+
+    protected static void setFileAsOrphan(SQLiteDatabase db, String filename) {
+        db.beginTransaction();
+        try {
+            ContentValues cv = new ContentValues();
+            cv.put(DatabaseHelper.FILE_COL, filename);
+            db.insertOrThrow(DbUtil.orphanFileTableName, DatabaseHelper.FILE_COL, cv);
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    protected static void unsetFileAsOrphan(SQLiteDatabase db, String filename) {
+        int deleteCount = db.delete(DbUtil.orphanFileTableName, DatabaseHelper.FILE_COL + "=?", new String[]{filename});
+        if (deleteCount != 1) {
+            Logger.log(AndroidLogger.SOFT_ASSERT,
+                    "Unable to unset orphaned file: " + deleteCount + " entries effected.b");
+        }
+    }
+
+    /**
+     * Remove files in the orphaned file table. Files are added to this table
+     * when file-backed db transactions fail, leaving the file on the
+     * filesystem.
+     *
+     * Order of operations expects filenames to be globally unique.
+     */
+    public static void removeOrphanedFiles(SQLiteDatabase db) {
+        Cursor cur = db.query(DbUtil.orphanFileTableName, new String[] {DatabaseHelper.FILE_COL}, null, null, null, null, null);
+        ArrayList<String> files = new ArrayList<>();
+        try {
+            if (cur.getCount() > 0) {
+                cur.moveToFirst();
+                int fileColIndex = cur.getColumnIndexOrThrow(DatabaseHelper.FILE_COL);
+                while (!cur.isAfterLast()) {
+                    files.add(cur.getString(fileColIndex));
+                    cur.moveToNext();
+                }
+            }
+        } finally {
+            if (cur != null) {
+                cur.close();
+            }
+        }
+
+        removeFiles(files);
+
+        db.beginTransaction();
+        try {
+            db.delete(DbUtil.orphanFileTableName, null, null);
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    protected static File newFileForEntry(File dbDir) throws IOException {
+        File newFile = getUniqueFilename(dbDir);
+
+        if (!newFile.createNewFile()) {
+            throw new RuntimeException("Trying to create a new file using existing filename; " +
+                    "Shouldn't be possible since we already checked for uniqueness");
+        }
+
+        return newFile;
+    }
+
+    private static File getUniqueFilename(File dbDir) {
+        String filename = PropertyUtils.genGUID(14);
```

Comment:
Sorry to keep being a pain in the ass about this. 14 chars is only about 72 bits of entropy, we wanna shoot for 128 as a general standard. There's actually a "generateUUID()" method specifically in the same class that will generate UUID v4 id's that I try to use for anything like this just to eliminate any concern.